### PR TITLE
Define TokenType, Ast node, statements, literals

### PR DESCRIPTION
**Pull request by zeroFruit**
resolved: #21, #22

details:

define TokenType, AST node, statement, expression, literal

### CODE REVIEW

**Comment by hihiboss** [comment]
There is an import cycle error. `ast` imported `parse` and `parse` imported `ast`.

**Comment by zeroFruit** [comment]
I solved cycle dependency problem by moving `ast.go` to `parse` package

**Comment by junbeomlee** [comment]
We should think about this unused `statementNode()` function and `Node interface{}`
And i don't think the token is necessary!! This struct is not used at all in the book also.
```
type ReturnStatement struct {
 	Token <- this token
 	ReturnValue Expression
 }
```
So let's talk about this.

**Review by junbeomlee — CHANGES_REQUESTED** [review]
We should think about this unused `statementNode()` function and `Node interface{}`
And i don't think the token is necessary!! This struct is not used at all in the book also.
```
type ReturnStatement struct {
 	Token <- this token
 	ReturnValue Expression
 }
```
So let's talk about this.

**Comment by zeroFruit** [comment]
I agree with removing `Token` so fixed still keep `Node` interface and deeply consider about method names. and named Statement node for `do()` because statements basically **do something**, and Expression node for `produce()` because expression basically **produce at least one value**. Referenced from here:

https://stackoverflow.com/a/4730559/6433772

**Comment by hihiboss** [comment]
~If `ast.go` is moved to `parse` package, is a dependency occurs?
Because, compiler component also uses `ast.go`...
Let's talk about whether moving `ast.go` is fine or not.~

**Comment by zeroFruit** [comment]
@hihiboss @junbeomlee refactored!

File: parse/token.go [hunk 1]
```diff
@@ -33,6 +33,36 @@ const (
 	Illegal TokenType = iota
 
 	// Identifiers + literals
-	Ident // add, foobar, x, y, ...
-	Int   // 1343456
+	Ident  // add, foobar, x, y, ...
+	Int    // 1343456
+	String // "hello world"
+	Bool   // true, false
+
```
Comment (junbeomlee) [inline parse/token.go:40]: Int    // 1343456
String // "hello world"
Bool   // true, false

Also we need int type tokenType

**Comment by zeroFruit** [comment]
Separated Token type with AST type (operator type, datastructure type) then compiler now only needs to know about AST type not Token type!

And parser needs to map Token type with AST type when building AST. Now the dependency is:
```
lexer -> Token
parser -> Token, AST
compiler -> AST
```
We can now remove dependency between Token and AST

**Review by junbeomlee — APPROVED** [review]
LGTM

**Comment by hihiboss** [comment]
It’s explicit!

**Review by hihiboss — APPROVED** [review]

**Pull request merged**